include/pmacc/particles/memory/frames/Frame.hpp

```diff
@@ -34,49 +34,149 @@
 #include "pmacc/traits/HasIdentifier.hpp"
 #include "pmacc/types.hpp"
 
+#include <boost/mp11.hpp>
+#include <boost/mp11/mpl.hpp>
 #include <boost/mpl/contains.hpp>
 #include <boost/mpl/find.hpp>
 #include <boost/mpl/list.hpp>
 #include <boost/mpl/map.hpp>
 #include <boost/utility/result_of.hpp>
 
+#include <llama/llama.hpp>
+
 namespace pmacc
 {
     namespace pmath = pmacc::math;
 
+    namespace detail
+    {
+        template<typename PICValueType>
+        using MakeLlamaField = llama::Field<PICValueType, typename traits::Resolve<PICValueType>::type::type>;
+
+        template<typename ValueTypeSeq>
+        using ValueTypeSeqToList =
+            typename boost::mpl::copy<ValueTypeSeq, boost::mpl::back_inserter<boost::mp11::mp_list<>>>::type;
+
+        template<typename ValueTypeSeq>
+        using RecordDimFromValueTypeSeq = boost::mp11::
+            mp_rename<boost::mp11::mp_transform<MakeLlamaField, ValueTypeSeqToList<ValueTypeSeq>>, llama::Record>;
+
+        template<std::size_t T_size, typename T_ParticleDescription, typename T_MemoryLayout>
+        struct ViewHolder
+        {
+            using IndexType = int; // TODO(bgruber): where do I get this type from?
```

Comment:
What index type should be used for all address arithmetic inside a `Frame`?

Reply:
`int` is fine, a frame will not hit the 2 gig element limit.